File: src/pixel_art_mcp/pixel_art.py
"""Native-grid authoring shared by authoring scripts and the exporter (stdlib only)."""

import json
import re
from typing import Any


class Canvas:
    """Integer pixels, top-left origin. A dot is transparent, never an eraser."""

    def __init__(self, width: int, height: int):
        if not 1 <= width <= 512 or not 1 <= height <= 512:
            raise ValueError("Canvas dimensions must be in 1..512")
        self.width, self.height = width, height
        self.pixels = [["."] * width for _ in range(height)]

    @classmethod
    def from_rows(cls, rows: list[str]) -> "Canvas":
        if not rows or not rows[0]:
            raise ValueError("Pixel rows must form a nonempty rectangle; row 0 must not be empty")
        bad = [(i, len(row)) for i, row in enumerate(rows) if len(row) != len(rows[0])]
        if bad:
            details = "; ".join(f"row {i}: expected {len(rows[0])}, actual {n}" for i, n in bad[:8])
            remaining = f"; {len(bad) - 8} more mismatched rows" if len(bad) > 8 else ""
            raise ValueError(
                f"Pixel rows must form a nonempty rectangle. {details}{remaining}. "
                "Row indices are zero-based; use numeric drawing commands for long shapes."
            )
        canvas = cls(len(rows[0]), len(rows))
        canvas.pixels = [list(row) for row in rows]
        return canvas

    @property
    def rows(self) -> list[str]:
        return ["".join(row) for row in self.pixels]

    def rect(self, x: int, y: int, width: int, height: int, color: str) -> "Canvas":
        if len(color) != 1 or min(x, y, width, height) < 0:
            raise ValueError("Use a palette symbol and nonnegative integer rectangle")
        if x + width > self.width or y + height > self.height:
            raise ValueError("Rectangle exceeds canvas")
        for row in self.pixels[y : y + height]:
            row[x : x + width] = [color] * width
        return self

    def stamp(self, x: int, y: int, rows: list[str]) -> "Canvas":
        patch = Canvas.from_rows(rows)
        if min(x, y) < 0 or x + patch.width > self.width or y + patch.height > self.height:
            raise ValueError("Stamp exceeds canvas")
        for dy, row in enumerate(patch.pixels):
            for dx, color in enumerate(row):
                if color != ".":
                    self.pixels[y + dy][x + dx] = color
        return self

    def mirrored(self) -> "Canvas":
        return Canvas.from_rows([row[::-1] for row in self.rows])

    def line(self, x1: int, y1: int, x2: int, y2: int, color: str) -> "Canvas":
        if len(color) != 1 or any(type(v) is not int for v in (x1, y1, x2, y2)):
            raise ValueError("Use integer line coordinates and a palette symbol")
        if not (
            0 <= x1 < self.width
            and 0 <= x2 < self.width
            and 0 <= y1 < self.height
            and 0 <= y2 < self.height
        ):
            raise ValueError("Line exceeds canvas")
        # Integer Bresenham: inclusive endpoints, no fractional coverage or antialiasing.
        dx, dy = abs(x2 - x1), -abs(y2 - y1)
        sx, sy = (1 if x1 < x2 else -1), (1 if y1 < y2 else -1)
        error = dx + dy
        while True:
            self.pixels[y1][x1] = color
            if (x1, y1) == (x2, y2):
                return self
            twice = 2 * error
            if twice >= dy:
                error += dy
                x1 += sx
            if twice <= dx:
                error += dx
                y1 += sy


class PixelArt:
    """Named, ordered layers with explicit per-view/per-frame pixel poses.

    Every layer draws the complete sprite through pixel helpers: these are
    finishing layers, not depth-tested textures, so author only the visible views.
    """

    def __init__(self, palette: dict[str, str], views: dict[int, tuple[int, int]]):
        if not 2 <= len(palette) <= 64 or any(
            len(k) != 1 or k == "." or not re.fullmatch(r"#[0-9a-fA-F]{6}", v)
            for k, v in palette.items()
        ):
            raise ValueError("Use 2..64 single-symbol #rrggbb palette entries; reserve '.'")
        if len({v.lower() for v in palette.values()}) != len(palette):
            raise ValueError("Palette colors must be distinct")
        if not views or any(
            angle not in (0, 90, 180, 270)
            or len(size) != 2
            or any(type(v) is not int or not 1 <= v <= 512 for v in size)
            for angle, size in views.items()
        ):
            raise ValueError("Declare each consumer view and its native canvas")
        self.palette = {k: v.lower() for k, v in palette.items()}
        self.views = {str(k): list(v) for k, v in views.items()}
        self.layers: list[dict[str, Any]] = []

    def layer(
        self,
        name: str,
        angle: int,
        canvas: Canvas,
        *,
        x: int = 0,
        y: int = 0,
        frame: int | None = None,
        min_pixels: int = 0,
        connected: bool = False,
    ) -> "PixelArt":
        if not name or len(name) > 100 or str(angle) not in self.views:
            raise ValueError("Layer needs a name and a declared view")
        if any(type(v) is not int for v in (x, y, min_pixels)) or min_pixels < 0:
            raise ValueError("Offsets and pixel budgets must be integers")
        if frame is not None and (type(frame) is not int or not 0 <= frame <= 1_000_000):
            raise ValueError("Invalid pose frame")
        if any(c not in self.palette and c != "." for row in canvas.rows for c in row):
            raise ValueError("Unknown palette symbol in pixel layer")
        layer = next((item for item in self.layers if item["name"] == name), None)
        if layer is None:
            layer = {"name": name, "poses": []}
            self.layers.append(layer)
        pose = {
            "angle": angle,
            "frame": frame,
            "rows": canvas.rows,
            "x": x,
            "y": y,
            "min_pixels": min_pixels,
            "connected": connected,
        }
        layer["poses"] = [
            p for p in layer["poses"] if (p["angle"], p["frame"]) != (angle, frame)
        ] + [pose]
        return self

    def poses(self, angle: int, frame: int) -> list[dict[str, Any]]:
        selected = []
        for layer in self.layers:
            choices = {p["frame"]: p for p in layer["poses"] if p["angle"] == angle}
            pose = choices.get(frame, choices.get(None))
            if pose is not None:
                selected.append({"name": layer["name"], **pose})
        return selected

    def to_dict(self) -> dict[str, Any]:
        return {
            "version": 1,
            "palette": self.palette,
            "views": self.views,
            "layers": self.layers,
        }

    def validate_target(
        self,
        layouts: list[dict[str, Any]],
        frames: list[int],
        spec: dict[str, Any],
    ) -> None:
        expected = {str(v["angle"]): [v["width"], v["height"]] for v in layouts}
        if self.views != expected:
            raise ValueError(f"Pixel views must match the configured canvases: {expected}")
        if spec["outline"]:
            raise ValueError("Draw outlines in pixel rows; configure_asset.outline must be false")
        if len(self.palette) > spec["colors"]:
            raise ValueError("Pixel palette exceeds configure_asset.colors")
        if spec.get("palette") and {v.lower() for v in spec["palette"]} != set(
            self.palette.values()
        ):
            raise ValueError("Pixel palette must match configure_asset.palette")
        if not self.layers:
            raise ValueError("A pixel-art definition with named layers is required")
        if not any(
            c != "."
            for layer in self.layers
            for p in layer["poses"]
            for row in p["rows"]
            for c in row
        ):
            raise ValueError("Pixel layers must contain authored ink")
        for layer in self.layers:
            for pose in layer["poses"]:
                w, h = expected[str(pose["angle"])]
                if len(pose["rows"]) > h or len(pose["rows"][0]) > w:
                    raise ValueError(f"Patch in {layer['name']!r} exceeds its view dimensions")
        for layout in layouts:
            for frame in frames:
                poses = self.poses(layout["angle"], frame)
                if not poses:
                    raise ValueError(
                        f"Missing pixel pose at angle {layout['angle']}, frame {frame}; add a "
                        f"default or exact-frame pose"
                    )
                ink = any(
                    c != "."
                    and 0 <= p["x"] + x < layout["width"]
                    and 0 <= p["y"] + y < layout["height"]
                    for p in poses
                    for y, row in enumerate(p["rows"])
                    for x, c in enumerate(row)
                )
                if not ink:
                    raise ValueError(f"Empty pose at angle {layout['angle']}, frame {frame}")

    @classmethod
    def from_dict(cls, data: dict[str, Any]) -> "PixelArt":
        if data.get("version") != 1:
            raise ValueError("Unsupported pixel art version")
        art = cls(data["palette"], {int(k): tuple(v) for k, v in data["views"].items()})
        for layer in data["layers"]:
            for pose in layer["poses"]:
                values = {k: v for k, v in pose.items() if k != "rows"}
                art.layer(layer["name"], canvas=Canvas.from_rows(pose["rows"]), **values)
        return art

    def save(self, scene: Any) -> None:
        # Round-trip validation also catches accidental direct edits to the layer data.
        data = self.from_dict(self.to_dict()).to_dict()
        scene["pixel_art"] = json.dumps(data, separators=(",", ":"))

    @classmethod
    def load(cls, scene: Any) -> "PixelArt":
        return cls.from_dict(json.loads(scene["pixel_art"]))

File: packages/pixel-core/src/pixel-art.ts
/**
 * Named, ordered layers with explicit per-view/per-frame pixel poses.
 *
 * Every layer draws the complete sprite through pixel helpers: these are finishing layers, not
 * depth-tested textures, so author only the visible views.
 *
 * Verbatim port of `PixelArt` in `src/pixel_art_mcp/pixel_art.py`. Every thrown `Error`'s
 * message below reproduces the source `ValueError`'s message exactly (see
 * `docs/typescript-rewrite.md`, Phase 3).
 */

import { Canvas } from "./canvas.js";
import type {
  AssetLayout,
  AssetValidationSpec,
  Layer,
  LayerOptions,
  PaletteInput,
  PixelArtDict,
  Pose,
  ResolvedPose,
  Scene,
  ViewsInput,
} from "./types.js";

const VALID_ANGLES = new Set([0, 90, 180, 270]);
const HEX_COLOR = /^#[0-9a-fA-F]{6}$/;

/**
 * Python `repr()` of a `str`, close enough for this module's use (layer names in error
 * messages): picks `'...'` unless the string holds a `'` but no `"`, and escapes `\`, quotes,
 * and the common whitespace/control characters. Real layer names are short authored identifiers
 * (see the examples in `tests/unit/test_pixel_art.py`), so this intentionally doesn't attempt
 * CPython's full non-ASCII "is this codepoint printable" logic -- flagged in this package's
 * final report as an edge case that isn't practical to reproduce byte-for-byte.
 */
function pyRepr(value: string): string {
  const quote = value.includes("'") && !value.includes('"') ? '"' : "'";
  let out = quote;
  for (const ch of value) {
    if (ch === "\\") out += "\\\\";
    else if (ch === quote) out += `\\${quote}`;
    else if (ch === "\n") out += "\\n";
    else if (ch === "\r") out += "\\r";
    else if (ch === "\t") out += "\\t";
    else {
      const code = ch.codePointAt(0) ?? 0;
      out += code < 0x20 || code === 0x7f ? `\\x${code.toString(16).padStart(2, "0")}` : ch;
    }
  }
  return out + quote;
}

/**
 * Builds `[key, value]` pairs the way a Python dict comprehension would: first-occurrence order,
 * last-value-wins, re-assigning an existing key does *not* move its position. `Map.set` already
 * has exactly this behavior, so this is just a thin, explicitly-documented wrapper -- used
 * instead of a plain JS object because object keys that look like array indices (all of this
 * package's angle keys: "0", "90", "180", "270") get silently reordered into ascending numeric
 * order by the JS engine regardless of insertion order, which a plain `Record`-based `expected`
 * dict would have made this module's error-message ordering diverge from Python's for
 * out-of-order `layouts` input.
 */
function orderedEntries<V>(pairs: readonly (readonly [string, V])[]): [string, V][] {
  const map = new Map<string, V>();
  for (const [k, v] of pairs) map.set(k, v);
  return [...map.entries()];
}

/** Python dict-repr of `{'<angle>': [width, height], ...}`, e.g. `{'0': [16, 16], '90': [32, 8]}`. */
function pyDictRepr(entries: readonly (readonly [string, [number, number]])[]): string {
  const parts = entries.map(([k, size]) => `'${k}': [${size[0]}, ${size[1]}]`);
  return `{${parts.join(", ")}}`;
}

/** Order-independent equality between `this.views` and a `validateTarget`-computed `expected`. */
function viewsEqual(
  a: Record<string, [number, number]>,
  bEntries: readonly (readonly [string, [number, number]])[],
): boolean {
  const aEntries = Object.entries(a);
  if (aEntries.length !== bEntries.length) return false;
  const bMap = new Map(bEntries);
  return aEntries.every(([k, v]) => {
    const bv = bMap.get(k);
    return bv?.[0] === v[0] && bv[1] === v[1];
  });
}

function setsEqual(a: ReadonlySet<string>, b: ReadonlySet<string>): boolean {
  if (a.size !== b.size) return false;
  for (const v of a) if (!b.has(v)) return false;
  return true;
}

export class PixelArt {
  palette: Record<string, string>;
  views: Record<string, [number, number]>;
  layers: Layer[];

  constructor(palette: PaletteInput, views: ViewsInput) {
    const paletteEntries = Object.entries(palette);
    const validPalette =
      paletteEntries.length >= 2 &&
      paletteEntries.length <= 64 &&
      paletteEntries.every(([k, v]) => k.length === 1 && k !== "." && HEX_COLOR.test(v));
    if (!validPalette) {
      throw new Error("Use 2..64 single-symbol #rrggbb palette entries; reserve '.'");
    }
    const loweredPalette = paletteEntries.map(([k, v]): [string, string] => [k, v.toLowerCase()]);
    if (new Set(loweredPalette.map(([, v]) => v)).size !== loweredPalette.length) {
      throw new Error("Palette colors must be distinct");
    }

    const viewEntries = Object.entries(views);
    const validViews =
      viewEntries.length > 0 &&
      viewEntries.every(([angleKey, size]) => {
        const angle = Number(angleKey);
        return (
          VALID_ANGLES.has(angle) &&
          size.length === 2 &&
          size.every((v) => Number.isInteger(v) && v >= 1 && v <= 512)
        );
      });
    if (!validViews) {
      throw new Error("Declare each consumer view and its native canvas");
    }

    this.palette = Object.fromEntries(loweredPalette);
    this.views = Object.fromEntries(
      viewEntries.map(([angleKey, size]): [string, [number, number]] => {
        const w = size[0];
        const h = size[1];
        if (w === undefined || h === undefined) {
          // Unreachable: `validViews` above already confirmed `size.length === 2`.
          throw new Error("Declare each consumer view and its native canvas");
        }
        return [angleKey, [w, h]];
      }),
    );
    this.layers = [];
  }

  layer(name: string, angle: number, canvas: Canvas, options: LayerOptions = {}): this {
    const { x = 0, y = 0, frame = null, min_pixels = 0, connected = false } = options;
    if (!name || name.length > 100 || !(String(angle) in this.views)) {
      throw new Error("Layer needs a name and a declared view");
    }
    if (
      !Number.isInteger(x) ||
      !Number.isInteger(y) ||
      !Number.isInteger(min_pixels) ||
      min_pixels < 0
    ) {
      throw new Error("Offsets and pixel budgets must be integers");
    }
    if (frame !== null && (!Number.isInteger(frame) || frame < 0 || frame > 1_000_000)) {
      throw new Error("Invalid pose frame");
    }
    for (const row of canvas.rows) {
      for (const c of row) {
        if (c !== "." && !(c in this.palette)) {
          throw new Error("Unknown palette symbol in pixel layer");
        }
      }
    }
    let layer = this.layers.find((item) => item.name === name);
    if (!layer) {
      layer = { name, poses: [] };
      this.layers.push(layer);
    }
    const pose: Pose = { angle, frame, rows: canvas.rows, x, y, min_pixels, connected };
    layer.poses = layer.poses.filter((p) => !(p.angle === angle && p.frame === frame));
    layer.poses.push(pose);
    return this;
  }

  poses(angle: number, frame: number): ResolvedPose[] {
    const selected: ResolvedPose[] = [];
    for (const layer of this.layers) {
      const choices = new Map<number | null, Pose>();
      for (const p of layer.poses) {
        if (p.angle === angle) choices.set(p.frame, p);
      }
      // Mirrors Python's `choices.get(frame, choices.get(None))`: exact-frame pose if one was
      // authored for this frame, else the layer's default (frame-less) pose, else nothing.
      const pose = choices.get(frame) ?? choices.get(null);
      if (pose) {
        selected.push({ name: layer.name, ...pose });
      }
    }
    return selected;
  }

  toDict(): PixelArtDict {
    return {
      version: 1,
      palette: { ...this.palette },
      views: { ...this.views },
      layers: this.layers.map((layer) => ({
        name: layer.name,
        poses: layer.poses.map((pose) => ({ ...pose, rows: [...pose.rows] })),
      })),
    };
  }

  validateTarget(
    layouts: readonly AssetLayout[],
    frames: readonly number[],
    spec: AssetValidationSpec,
  ): void {
    const expectedEntries = orderedEntries(
      layouts.map((v): [string, [number, number]] => [String(v.angle), [v.width, v.height]]),
    );
    if (!viewsEqual(this.views, expectedEntries)) {
      throw new Error(
        `Pixel views must match the configured canvases: ${pyDictRepr(expectedEntries)}`,
      );
    }
    if (spec.outline) {
      throw new Error("Draw outlines in pixel rows; configure_asset.outline must be false");
    }
    if (Object.keys(this.palette).length > spec.colors) {
      throw new Error("Pixel palette exceeds configure_asset.colors");
    }
    if (spec.palette && spec.palette.length > 0) {
      const specColors = new Set(spec.palette.map((c) => c.toLowerCase()));
      const ownColors = new Set(Object.values(this.palette));
      if (!setsEqual(specColors, ownColors)) {
        throw new Error("Pixel palette must match configure_asset.palette");
      }
    }
    if (this.layers.length === 0) {
      throw new Error("A pixel-art definition with named layers is required");
    }
    const hasInk = this.layers.some((layer) =>
      layer.poses.some((pose) => pose.rows.some((row) => Array.from(row).some((c) => c !== "."))),
    );
    if (!hasInk) {
      throw new Error("Pixel layers must contain authored ink");
    }
    const expected = new Map(expectedEntries);
    for (const layer of this.layers) {
      for (const pose of layer.poses) {
        const size = expected.get(String(pose.angle));
        if (!size) continue; // Unreachable: the views-equality check above guarantees a match.
        const [w, h] = size;
        const firstRow = pose.rows[0];
        if (pose.rows.length > h || (firstRow !== undefined && firstRow.length > w)) {
          throw new Error(`Patch in ${pyRepr(layer.name)} exceeds its view dimensions`);
        }
      }
    }
    for (const layout of layouts) {
      for (const frame of frames) {
        const poses = this.poses(layout.angle, frame);
        if (poses.length === 0) {
          throw new Error(
            `Missing pixel pose at angle ${layout.angle}, frame ${frame}; add a default or ` +
              "exact-frame pose",
          );
        }
        const hasInkInBounds = poses.some((p) =>
          p.rows.some((row, dy) =>
            Array.from(row).some((c, dx) => {
              if (c === ".") return false;
              const px = p.x + dx;
              const py = p.y + dy;
              return px >= 0 && px < layout.width && py >= 0 && py < layout.height;
            }),
          ),
        );
        if (!hasInkInBounds) {
          throw new Error(`Empty pose at angle ${layout.angle}, frame ${frame}`);
        }
      }
    }
  }

  /**
   * Re-validates everything by replaying every stored pose through `layer()` -- the core
   * correctness guarantee, matching Python's `from_dict` exactly. Like the Python source, this
   * trusts `data` to already have the right shape beyond the `version` check: a malformed object
   * (missing `palette`/`views`/`layers`/pose keys) fails with a generic JS `TypeError` reading an
   * undefined property, the same way Python's version fails with a generic `KeyError` -- neither
   * is a domain `ValueError`/`Error` with a curated message, so this intentionally does not add
   * extra validation beyond what the source has (see this package's final report).
   */
  static fromDict(data: PixelArtDict): PixelArt {
    if (data.version !== 1) {
      throw new Error("Unsupported pixel art version");
    }
    const views: ViewsInput = Object.fromEntries(
      Object.entries(data.views).map(([k, v]): [number, readonly number[]] => [Number(k), v]),
    );
    const art = new PixelArt(data.palette, views);
    for (const layer of data.layers) {
      for (const pose of layer.poses) {
        art.layer(layer.name, pose.angle, Canvas.fromRows(pose.rows), {
          x: pose.x,
          y: pose.y,
          frame: pose.frame,
          min_pixels: pose.min_pixels,
          connected: pose.connected,
        });
      }
    }
    return art;
  }

  save(scene: Scene): void {
    // Round-trip validation also catches accidental direct edits to the layer data.
    const data = PixelArt.fromDict(this.toDict()).toDict();
    scene["pixel_art"] = JSON.stringify(data);
  }

  static load(scene: Scene): PixelArt {
    // The one unavoidable cast in this package: `Scene` is deliberately `Record<string,
    // unknown>` (see types.ts), so extracting its one field `pixel-core` cares about needs an
    // explicit assertion to hand a `string` to `JSON.parse`. A wrong shape here fails the same
    // loose way Python's `load` does (see `fromDict`'s doc comment) -- a generic runtime error,
    // not a curated one.
    return PixelArt.fromDict(JSON.parse(scene["pixel_art"] as string) as PixelArtDict);
  }
}

// Runtime protection (see docs/typescript-rewrite.md, "Core design decisions"): an untrusted-but-
// sandboxed render script must never be able to alter PixelArt's behavior for any other job.
// Freezing the constructor and its prototype means e.g. `PixelArt.prototype.layer = () => {}`
// throws a TypeError immediately, under ESM's implicit strict mode, instead of silently
// succeeding.
Object.freeze(PixelArt);
Object.freeze(PixelArt.prototype);
